### PR TITLE
[refactor cluster-task-manage 1/n] separate resource reporting logic into helper class

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

Separate Scheduler Resource Reporting logic into a separate class for better readability and maintainability.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scv119** [comment]
Also we think there are two bugs in the scheduler resource reporting logic:

1. The autoscaler expects that the [resource_demands](https://github.com/ray-project/ray/blob/master/src/ray/protobuf/gcs.proto#L388) is unique per resource shape. however the way scheduler resource reporting implemented didn't dedupe the resource shapes between `tasks_to_schedule_` and `tasks_to_dispatch_`.

2. We report client side pending requests from worker to local raylet (https://github.com/ray-project/ray/blob/master/src/ray/raylet/scheduling/cluster_task_manager.cc#L843); however in the case that the lease request is sent to a remote raylet (spilling, or locality aware scheduling), this might under count the resource requirement.

cc @wuisawesome  @jjyao

File: src/ray/raylet/scheduling/scheduler_resource_reporter.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#include "absl/container/flat_hash_map.h"
+#include "ray/common/ray_config.h"
+#include "ray/common/task/task_spec.h"
+#include "ray/raylet/scheduling/internal.h"
+
+namespace ray {
+namespace raylet {
+
+/// Helper class that reports resource_load and resource_load_by_shape to gcs.
+class SchedulerResourceReporter {
```
Comment (iycheng) [inline src/ray/raylet/scheduling/scheduler_resource_reporter.h:26]: Is this local scheduler or cluster?

**Review by fishbone — APPROVED** [review]

**Comment by iycheng** [comment]
Just want to make sure I understand it correctly, after this one,
src/ray/raylet/scheduling/scheduler_resource_reporter.h basically will be something like the Reporter concept we talked about before, right?

**Comment by jjyao** [comment]
> We report client side pending requests from worker to local raylet (https://github.com/ray-project/ray/blob/master/src/ray/raylet/scheduling/cluster_task_manager.cc#L843); however in the case that the lease request is sent to a remote raylet (spilling, or locality aware scheduling), this might under count the resource requirement.

We only report tasks without pending lease requests to local raylet. For tasks with pending lease requests, they will be reported by the raylets that receive the lease requests. Combined together, gcs should see an almost accurate resource requirement.

**Review by jjyao — APPROVED** [review]

File: src/ray/raylet/scheduling/scheduler_resource_reporter.h (hunk 1 above)
Comment (scv119) [reply src/ray/raylet/scheduling/scheduler_resource_reporter.h:26]: yeah this is both

**Comment by wuisawesome** [comment]
> The autoscaler expects that the [resource_demands](https://github.com/ray-project/ray/blob/master/src/ray/protobuf/gcs.proto#L388) is unique per resource shape. however the way scheduler resource reporting implemented didn't dedupe the resource shapes between tasks_to_schedule_ and tasks_to_dispatch_.

Let's remove that comment. The code doesn't seem to make that assumption. https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/autoscaler/_private/monitor.py?L85

> We report client side pending requests from worker to local raylet (https://github.com/ray-project/ray/blob/master/src/ray/raylet/scheduling/cluster_task_manager.cc#L843); however in the case that the lease request is sent to a remote raylet (spilling, or locality aware scheduling), this might under count the resource requirement.

I think that line reference changed? But if you're referring to the worker backlog reporting, I think the leawse request will get counted, but at the other raylet? (there is a short race condition when it's in transit and not counted i guess).

**Pull request merged**